File: crates/viben-executors/src/executors/cursor.rs
//! Cursor Agent executor

use std::{
    path::{Path, PathBuf},
    sync::Arc,
};

use async_trait::async_trait;
use derivative::Derivative;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use ts_rs::TS;
use viben_utils::MsgStore;

use super::{
    AvailabilityInfo, ExecutorApprovalService, ExecutorError, SpawnedChild,
    StandardCodingAgentExecutor,
};
use crate::env::ExecutionEnv;

/// Cursor Agent executor
#[derive(Derivative, Clone, Serialize, Deserialize, TS, JsonSchema)]
#[derivative(Debug, PartialEq, Default)]
pub struct CursorAgent {
    /// Model to use
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
}

impl CursorAgent {
    /// Create a new Cursor agent executor
    pub fn new() -> Self {
        Self::default()
    }
}

#[async_trait]
impl StandardCodingAgentExecutor for CursorAgent {
    fn use_approvals(&mut self, _approvals: Arc<dyn ExecutorApprovalService>) {}

    async fn spawn(
        &self,
        _current_dir: &Path,
        _prompt: &str,
        _env: &ExecutionEnv,
    ) -> Result<SpawnedChild, ExecutorError> {
        // Cursor doesn't have a CLI - it requires the Cursor IDE
        Err(ExecutorError::AuthRequired(
            "Cursor requires the Cursor IDE to be running".to_string(),
        ))
    }

    async fn spawn_follow_up(
        &self,
        _current_dir: &Path,
        _prompt: &str,
        _session_id: &str,
        _reset_to_message_id: Option<&str>,
        _env: &ExecutionEnv,
    ) -> Result<SpawnedChild, ExecutorError> {
        Err(ExecutorError::FollowUpNotSupported(
            "Cursor does not support CLI follow-up sessions".to_string(),
        ))
    }

    fn normalize_logs(&self, _raw_logs: Arc<MsgStore>, _worktree_path: &Path) {}

    fn default_mcp_config_path(&self) -> Option<PathBuf> {
        // Cursor stores MCP config in its own settings
        #[cfg(target_os = "macos")]
        {
            dirs::home_dir().map(|home| {
                home.join("Library/Application Support/Cursor/User/globalStorage/mcp.json")
            })
        }
        #[cfg(target_os = "linux")]
        {
            dirs::config_dir().map(|config| config.join("Cursor/User/globalStorage/mcp.json"))
        }
        #[cfg(target_os = "windows")]
        {
            dirs::config_dir().map(|config| config.join("Cursor/User/globalStorage/mcp.json"))
        }
    }

    fn get_availability_info(&self) -> AvailabilityInfo {
        // Check if Cursor is installed by looking for its config directory
        let cursor_path = self.default_mcp_config_path();
        if let Some(path) = cursor_path {
            if path.parent().map(|p| p.exists()).unwrap_or(false) {
                return AvailabilityInfo::InstallationFound;
            }
        }
        AvailabilityInfo::NotFound
    }
}

File: packages/core/src/executors/executors/cursor.ts
/**
 * Cursor Agent executor
 */
import { spawn } from "node:child_process";
import { existsSync } from "node:fs";
import { join } from "node:path";
import { CommandBuilder } from "../command";
import { getConfigDir } from "../utils";
import type {
  ExecutorConfig,
  SpawnedChild,
  ExecutionEnv,
  StandardCodingAgentExecutor,
  AgentCapability,
  AvailabilityInfo,
} from "../types";
import { ExecutorError } from "../../error";
import { which } from "../utils";

const BASE_COMMAND = "cursor";

/**
 * Cursor Agent executor configuration
 */
export interface CursorAgentConfig extends ExecutorConfig {
  /** Model to use */
  model?: string;
}

/**
 * Cursor Agent executor
 */
export class CursorAgent implements StandardCodingAgentExecutor {
  readonly type = "CURSOR_AGENT" as const;
  private config: CursorAgentConfig;

  constructor(config: CursorAgentConfig = {}) {
    this.config = config;
  }

  async spawn(
    currentDir: string,
    prompt: string,
    env: ExecutionEnv
  ): Promise<SpawnedChild> {
    let builder = CommandBuilder.new(this.config.baseCommandOverride || BASE_COMMAND)
      .addParams("--prompt", prompt);

    if (this.config.model) {
      builder = builder.addParams("--model", this.config.model);
    }

    const commandParts = builder.buildInitial();

    const programPath = await which(commandParts.program);
    if (!programPath) {
      throw ExecutorError.executableNotFound(commandParts.program);
    }

    const spawnEnv = {
      ...process.env,
      ...env.vars,
      ...commandParts.env,
      ...(this.config.env || {}),
    };

    const child = spawn(programPath, commandParts.args, {
      cwd: currentDir,
      env: spawnEnv,
      stdio: ["pipe", "pipe", "pipe"],
    });

    return { child };
  }

  async spawnFollowUp(
    _currentDir: string,
    _prompt: string,
    _sessionId: string,
    _resetToMessageId: string | undefined,
    _env: ExecutionEnv
  ): Promise<SpawnedChild> {
    throw ExecutorError.followUpNotSupported("CursorAgent");
  }

  defaultMcpConfigPath(): string | null {
    return join(getConfigDir(), "Cursor", "User", "settings.json");
  }

  getAvailabilityInfo(): AvailabilityInfo {
    const configPath = this.defaultMcpConfigPath();
    if (configPath && existsSync(configPath)) {
      return { status: "INSTALLATION_FOUND" };
    }
    return { status: "NOT_FOUND" };
  }

  capabilities(): AgentCapability[] {
    return ["SETUP_HELPER"];
  }
}

export function createCursorAgent(config?: CursorAgentConfig): CursorAgent {
  return new CursorAgent(config);
}
